### PR TITLE
15175 - Changes for search for SBC-AUTH.

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #, if available:* bcgov/entity/issues/15175

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: api/namex/models/request.py [hunk 1]
```diff
@@ -88,11 +88,11 @@ class Request(db.Model):
     activeUser = db.relationship('User', backref=backref('active_user', uselist=False), foreign_keys=[userId])
     submitter = db.relationship('User', backref=backref('submitter', uselist=False), foreign_keys=[submitter_userid])
     # Relationships - Names
-    names = db.relationship('Name', lazy='dynamic')
+    names = db.relationship('Name', lazy='select')
```
Comment (seeker25) [inline api/namex/models/request.py:91]: https://github.com/pallets-eco/flask-sqlalchemy/pull/459

Can't override `lazy='dynamic'` unfortunately

File: api/namex/resources/requests.py [hunk 2]
```diff
@@ -383,6 +378,65 @@ def post(self, *args, **kwargs):
...
+        name = search.get('name', None)
+        identifiers = search.get('identifiers', [])
+        search_identifier = search.get('searchIdentifier', None)
+        status = search.get('status', None)
+        page = search.get('page', 1)
+        limit = search.get('limit', 20)
+        q = RequestDAO.query.filter(RequestDAO.nrNum.in_(identifiers)) \
+            .options(
+                lazyload('*'),
+                eagerload(RequestDAO.names).load_only(Name.state, Name.name),
```
Comment (seeker25) [inline api/namex/resources/requests.py:400]: I tried to load as few fields as possible. 

Example query after serialization:

```
 SELECT requests.nr_num AS requests_nr_num, requests.request_type_cd AS requests_request_type_cd, requests.nature_business_info AS requests_nature_business_info, requests.state_cd AS requests_state_cd, requests.entity_type_cd AS requests_entity_type_cd, requests.id AS requests_id, names_1.id AS names_1_id, names_1.name AS names_1_name, names_1.state AS names_1_state, applicants_1.party_id AS applicants_1_party_id, applicants_1.phone_number AS applicants_1_phone_number, applicants_1.email_address AS applicants_1_email_address 
FROM requests LEFT OUTER JOIN names AS names_1 ON requests.id = names_1.nr_id LEFT OUTER JOIN applicants AS applicants_1 ON requests.id = applicants_1.nr_id 
WHERE requests.nr_num IN (%(nr_num_1_1)s, %(nr_num_1_2)s, %(nr_num_1_3)s, %(nr_num_1_4)s, %(nr_num_1_5)s) AND (requests.name_search ILIKE %(name_search_1)s OR requests.name_search ILIKE %(name_search_2)s OR requests.name_search ILIKE %(name_search_3)s) AND requests.id IN (SELECT anon_1.id 
FROM (SELECT requests.id AS id 
FROM requests 
WHERE requests.nr_num IN (%(nr_num_1_1)s, %(nr_num_1_2)s, %(nr_num_1_3)s, %(nr_num_1_4)s, %(nr_num_1_5)s) AND (requests.name_search ILIKE %(name_search_1)s OR requests.name_search ILIKE %(name_search_2)s OR requests.name_search ILIKE %(name_search_3)s) GROUP BY requests.id 
 LIMIT %(param_1)s OFFSET %(param_2)s) AS anon_1)
```

There is an index on nr_id already that exists.

File: api/namex/resources/requests.py [hunk 3]
```diff
@@ -2,14 +2,17 @@
 
 TODO: Fill in a larger description once the API is defined for V1
 """
+from collections import namedtuple
```
Comment (seeker25) [inline api/namex/resources/requests.py:5]: Will remove

File: api/namex/resources/requests.py [hunk 4]
```diff
@@ -2,14 +2,17 @@
...
 from http import HTTPStatus
 from flask import request, jsonify, g, current_app, get_flashed_messages
 from flask_restx import Namespace, Resource, fields, cors
 from flask_jwt_oidc import AuthError
 
 from namex.constants import DATE_TIME_FORMAT_SQL
+from namex.models.request import RequestsAuthSearchSchema
 from namex.utils.logging import setup_logging
 
+from sqlalchemy.orm import load_only, joinedload, lazyload, subqueryload, selectinload, eagerload
```
Comment (seeker25) [inline api/namex/resources/requests.py:15]: Will fix, probably don't need all of these

File: api/tests/python/end_points/test_namex_search.py [hunk 5]
```diff
@@ -591,35 +591,44 @@ def test_namex_search_submitted_start_and_end_date_invalid_date_format(client,
...
+@pytest.mark.parametrize('identifiers, search_identifier, status, name, total_results', [
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], '', State.APPROVED, '', 1),
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], '', State.DRAFT, '', 1),
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], '', '', 'TEST', 5),
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], '', '', 'BAD', 0),
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], 'NR 1', '', '', 1),
+    ([], '', '', '', 0),
+    (['NR 0', 'NR 1', 'NR 2', 'NR 3', 'NR 4'], '', '', '', 5),
+])
+def test_namex_search_direct_nrs(client, jwt, app, identifiers, search_identifier, status, name, total_results):
```
Comment (kialj876) [inline api/tests/python/end_points/test_namex_search.py:603]: I find it pretty helpful to have the name of each test as the first arg in each param tuple. That way if one case fails its really easy for everyone to know the scenario they broke

**Review by kialj876 — APPROVED** [review]
Looks good to me

File: api/tests/python/end_points/test_namex_search.py (hunk 5 above)
Comment (kialj876) [reply api/tests/python/end_points/test_namex_search.py:603]: oops didn't mean to approve the draft haha

Comment (seeker25) [reply api/tests/python/end_points/test_namex_search.py:603]: Great suggestion, I'll fix this

File: api/namex/resources/requests.py [hunk 6]
```diff
@@ -383,6 +377,66 @@ def post(self, *args, **kwargs):
...
+@cors_preflight("POST")
+@api.route('/search', methods=['POST'])
+class RequestSearch(Resource):
+
+    @staticmethod
+    @cors.crossdomain(origin='*')
+    @jwt.has_one_of_roles([User.SYSTEM])
+    def post():
+        search = request.get_json()
+        name = search.get('name', None)
```
Comment (seeker25) [inline api/namex/resources/requests.py:390]: After talking with Kial, we need to axe `name`, `search_identifier`, `status`

**Review by kialj876 — APPROVED** [review]

**Review by pwei1018 — APPROVED** [review]

**Comment by seeker25** [comment]
Working on a fix for this PR.

**Pull request merged**